src/IO/FileEncryptionCommon.h

```diff
@@ -16,87 +16,82 @@ class WriteBuffer;
-    void inc() { ++counter; }
-    void inc(size_t n) { counter += n; }
-    void set(size_t n) { counter = n; }
-
-private:
-    UInt128 iv;
-    UInt128 counter = 0;
-    mutable String local;
-};
+    static constexpr const size_t kSize = 16;
 
+    InitVector() = default;
+    explicit InitVector(const UInt128 & counter_) { set(counter_); }
 
-class EncryptionKey
-{
-public:
-    EncryptionKey(const String & key_) : key(key_) { }
-    size_t size() const { return key.size(); }
-    const String & str() const { return key; }
+    void set(const UInt128 & counter_) { counter = counter_; }
+    UInt128 get() const { return counter; }
 
-private:
-    String key;
-};
+    void read(ReadBuffer & in);
+    void write(WriteBuffer & out) const;
 
+    /// Write 16 bytes of the counter to a string in big endian order.
+    /// We need big endian because the used cipher algorithms treat an initialization vector as a counter in big endian.
+    String toString() const;
 
-class Encryption
-{
-public:
-    Encryption(const String & iv_, const EncryptionKey & key_, size_t offset_);
+    /// Converts a string of 16 bytes length in big endian order to a counter.
+    static InitVector fromString(const String & str_);
 
-protected:
-    size_t blockOffset(size_t pos) const { return pos % block_size; }
-    size_t blocks(size_t pos) const { return pos / block_size; }
-    size_t partBlockSize(size_t size, size_t off) const;
-    const EVP_CIPHER * get() const { return evp_cipher; }
+    /// Adds a specified offset to the counter.
+    InitVector & operator++() { ++counter; return *this; }
+    InitVector operator++(int) { InitVector res = *this; ++counter; return res; }
+    InitVector & operator+=(size_t offset) { counter += offset; return *this; }
+    InitVector operator+(size_t offset) const { InitVector res = *this; return res += offset; }
 
-    const EVP_CIPHER * evp_cipher;
-    const String init_vector;
-    const EncryptionKey key;
-    size_t block_size;
-
-    /// absolute offset
-    size_t offset = 0;
-};
-
-
-class Encryptor : public Encryption
-{
-public:
-    using Encryption::Encryption;
-    void encrypt(const char * plaintext, WriteBuffer & buf, size_t size);
+    /// Generates a random initialization vector.
+    static InitVector random();
 
 private:
-    String encryptPartialBlock(const char * partial_block, size_t size, const InitVector & iv, size_t off) const;
-    String encryptNBytes(const char * data, size_t bytes, const InitVector & iv) const;
+    UInt128 counter = 0;
 };
 
 
-class Decryptor : public Encryption
+/// Encrypts or decrypts data.
+class Encryptor
 {
 public:
-    Decryptor(const String & iv_, const EncryptionKey & key_) : Encryption(iv_, key_, 0) { }
-    void decrypt(const char * ciphertext, char * buf, size_t size, size_t off);
+    /// The `key` should have length 128 or 192 or 256.
+    /// According to the key's length aes_128_ctr or aes_192_ctr or aes_256_ctr will be used for encryption.
```

Comment:
I'd rather force user to choose algorithm explicitly in config.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/26465